File: clojure/src/project_euler/sol/p0089.clj
(ns project-euler.sol.p0089
  (:require
   [clojure.string :as str]
   [project-euler.lib.util :as util]))

;;;   step 1:
;;;     IIIIIIIII     IX
;;;     XXXXXXXXX     XC
;;;     CCCCCCCCC     CM
;;;
;;;   step 2:
;;;     VIIII         IX
;;;     LXXXX         XC
;;;     DCCCC         CM
;;;
;;;   step 3:
;;;     IIIII         V
;;;     XXXXX         L
;;;     CCCCC         D
;;;
;;;   step 4:
;;;     IIII          IV
;;;     XXXX          XL
;;;     CCCC          CD
(defn- replace-roman-nums
  "Replace from long Roman numerals."
  [s]
  (-> s
      (str/replace #"IIIIIIIII|XXXXXXXXX|CCCCCCCCC" "##") ; IX/XC/CM
      (str/replace #"VIIII|LXXXX|DCCCC" "##")             ; IX/XC/CM
      (str/replace #"IIIII|XXXXX|CCCCC" "#")              ; V/L/D
      (str/replace #"IIII|XXXX|CCCC" "##"))) ; IV/XL/CD

(defn solve
  ([]
   (solve (util/read-data "0089_roman.txt")))
  ([data]
   (->> (map #(- (count %) (count (replace-roman-nums %))) data)
        (apply +))))

File: julia/Euler/src/bin/p0089.jl

# project euler: problem 89

#=
  step 1:
    IIIIIIIII     IX
    XXXXXXXXX     XC
    CCCCCCCCC     CM

  step 2:
    VIIII         IX
    LXXXX         XC
    DCCCC         CM

  step 3:
    IIIII         V
    XXXXX         L
    CCCCC         D

  step 4:
    IIII          IV
    XXXX          XL
    CCCC          CD
=#

module Prob0089

function replace_numbers(line)
    replace(line, r"IIIIIIIII|XXXXXXXXX|CCCCCCCCC" => "##",
                  r"VIIII|LXXXX|DCCCC" => "##",
                  r"IIIII|XXXXX|CCCCC" => "#",
                  r"IIII|XXXX|CCCC" => "##")
end

function solve_0089(fname = "0089_roman.txt")
    data = readlines(joinpath((@__DIR__), "../../assets", fname))

    acc = 0
    for line in data
        acc += length(line) - length(replace_numbers(line))
    end
    acc
end

end #module

using .Prob0089: solve_0089
export solve_0089
